### PR TITLE
Task#C Documentation Comments

### PR DESCRIPTION
**Pull request by pcsmartapps**
Completed documentation and comments for whole project per good coding standards established in class.

### CODE REVIEW

File: include/atomics/receiver_cadmium.hpp [hunk 1]
```diff
@@ -22,48 +33,71 @@
...
 template<typename TIME>
 class Receiver{
-    using defs = receiver_defs; // putting definitions in context
+    // putting definitions in context
+    using defs = receiver_defs; 
     public:
     //Parameters to be overwriten when instantiating the atomic model
-    TIME PREPARATION_TIME;
-    //default constructor
+    TIME PREPARATION_TIME; /**< PREPARATION_TIME is a constant that contains the time delay from acknowledge to output. */
```
Comment (p-sky) [inline include/atomics/receiver_cadmium.hpp]: Can include brief member comment here after detailed comment and in other places:
/**< Detailed comment. */ //!<Brief comment.

**Review by p-sky — CHANGES_REQUESTED** [review]
reviewed the following files:
receiver_cadmium.hpp
sender_cadmium.hpp
subnet_cadmium.hpp

File: src/data_structures/message.cpp [hunk 2]
```diff
@@ -1,25 +1,32 @@
...
-
-//updated relative path --Syed Omar
 #include "../../include/data_structures/message.hpp"
 
-/***************************************************/
-/************* Output stream ************************/
-/***************************************************/
-
+/**\brief Output stream
+ *
```
Comment (p-sky) [inline src/data_structures/message.cpp]: Can add detailed description or just have detailed without brief description

**Review by p-sky — CHANGES_REQUESTED** [review]

File: include/atomics/receiver_cadmium.hpp (hunk 1 above)
Comment (pcsmartapps) [reply include/atomics/receiver_cadmium.hpp]: Modified

File: src/data_structures/message.cpp (hunk 2 above)
Comment (pcsmartapps) [reply src/data_structures/message.cpp]: Done. Thanks for your informative feedback

File: README.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
+# New Document# **ALTERNATE BIT PROTOCOL SIMULATOR IMPLEMENTED IN CADMIUM**
```
Comment (shubhamagrawal6629) [inline README.md]: New Document is not needed before the title

File: README.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+   - receiver_input_test.txt
+   - receiver_test_output.txt
+   - sender_input_test_ack_In.txt
+   - sender_input_test_control_In.txt
+   - sender_test_output.txt
+   - subnet_input_test.txt
+   - subnet_test_output.txt
+3. include [This folder contains the header files used for unit testing]
+      will be updated in future
+#####
```
Comment (shubhamagrawal6629) [inline README.md]: no need of this hash

File: README.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+---
+## ORGANIZATION : CARLETON UNIVERSITY
+---
+## AUTHORS :
+### ORIGINALLY DEVELOPED BY : Cristina Ruiz Martin
+### MODIFIED BY : Syed Omar Azam, Mohamed Ali and Ammar Al-Mahbashi                  
+---
+## INTRODUCTION:
+---
+**ABP** (Alternating Bit Protocol) is a communication protocol to ensure reliable transmission through unreliable network. The sender sends a packet and waits for an acknowledgement. If the acknowledgement doesn't arrive within a predefined time, the sender re-sends this packet until it receives an expected acknowledgement and then sends the next packet. In order to distinguish two consecutive packets, the sender adds an additional bit on each packet (called alternating bit because the sender uses 0 and 1 alternatively). This software is a simulator to replicate the behavior of the Alternating Bit Protocol.
```
Comment (shubhamagrawal6629) [inline README.md]: It would be better if you use Alternation Bit Protocol(ABP) instead of ABP (Alternation Bit Protocol)

File: README.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+   - receiver_test_output.txt
+   - sender_input_test_ack_In.txt
+   - sender_input_test_control_In.txt
+   - sender_test_output.txt
+   - subnet_input_test.txt
+   - subnet_test_output.txt
+3. include [This folder contains the header files used for unit testing]
+      will be updated in future
+#####
+1. makefile
```
Comment (shubhamagrawal6629) [inline README.md]: it can be without number like other folders

File: README.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+---
+
+## STEPS TO RUN THE SIMULATOR
+---
+alternatebitprot.pdf contains the explanation of this simulator.
```
Comment (shubhamagrawal6629) [inline README.md]: you can mention where is the file located or give reference to it by [document](path to your master document folder)

**Review by shubhamagrawal6629 — CHANGES_REQUESTED** [review]

**Comment by shubhamagrawal6629** [comment]
Some suggestion for readme

1. You can provide table of contents with hyperlink
2. You can specify how to install cadmium and destimes either giving the steps or by referring the documents
3. You can also provide hyperlinks to author so that one can access their profile directly

**Comment by shubhamagrawal6629** [comment]
For your gitignore files it looks like it is not working as you still have all the exe in the bin folder

File: test/src/main_receiver.cpp [hunk 8]
```diff
@@ -26,36 +37,69 @@ using hclock = chrono::high_resolution_clock;
...
 struct outp : public cadmium::out_port<message_t> {};
 
 
-/********************************************/
-/****** APPLICATION GENERATOR *******************/
-/********************************************/
+/**
+ * This is application generator class takes file path
+ * parameter and waits for input
+ * @tParam message T
```
Comment (shubhamagrawal6629) [inline test/src/main_receiver.cpp]: it will be @param

File: test/src/main_receiver.cpp [hunk 9]
```diff
@@ -26,36 +37,69 @@ using hclock = chrono::high_resolution_clock;
...
+ */
 template<typename T>
 class ApplicationGen : public iestream_input<message_t,T> {
     public:
         ApplicationGen() = default;
-        ApplicationGen(const char* file_path) : iestream_input<message_t,T>(file_path) {}
+
+        /**
+         * A parameterized contructor for class application generator
```
Comment (shubhamagrawal6629) [inline test/src/main_receiver.cpp]: it will be constructor

Comment (shubhamagrawal6629) [reply test/src/main_receiver.cpp]: and contains

Comment (shubhamagrawal6629) [reply test/src/main_receiver.cpp]: there are more spelling mistakes in line 155 and 99 and 200

Comment (shubhamagrawal6629) [reply test/src/main_receiver.cpp]: there are more spelling mistakes in sender and subnet .cpp files

File: README.md (hunk 3 above)
Comment (syedomarazam) [reply README.md]: removed.

File: README.md (hunk 4 above)
Comment (syedomarazam) [reply README.md]: removed.

File: README.md (hunk 5 above)
Comment (syedomarazam) [reply README.md]: done. either way, it is ok.

File: README.md (hunk 7 above)
Comment (syedomarazam) [reply README.md]: implemented this functionality.

File: README.md (hunk 6 above)
Comment (syedomarazam) [reply README.md]: it is not a folder. but have implemented per your suggestion. thanks

File: test/src/main_receiver.cpp (hunk 8 above)
Comment (syedomarazam) [reply test/src/main_receiver.cpp]: fixed

File: test/src/main_receiver.cpp (hunk 9 above)
Comment (syedomarazam) [reply test/src/main_receiver.cpp]: fixed spelling mistakes in all the mentioned files.

**Review by syedomarazam — COMMENTED** [review]
Some suggestion for readme from Shubham:

You can provide a table of contents with a hyperlink - IMPLEMENTED
You can specify how to install Cadmium and DESTimes either giving the steps or by referring the documents - IMPLEMENTED
You can also provide hyperlinks to the author so that one can access their profile directly - IMPLEMENTED

**Comment by syedomarazam** [comment]
fixed all the changes requested by code reviewers and tested successfully.

File: README.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
+#**ALTERNATE BIT PROTOCOL SIMULATOR IMPLEMENTED IN CADMIUM**
```
Comment (shubhamagrawal6629) [inline README.md]: There is # before

**Review by shubhamagrawal6629 — CHANGES_REQUESTED** [review]

File: README.md (hunk 10 above)
Comment (syedomarazam) [reply README.md]: removed 👍

**Review by shubhamagrawal6629 — APPROVED** [review]
Project documentation (including README file) and code comments pull request is in sync with the requirements established for the project as part of deliverable C. Good job!

**Review by p-sky — APPROVED** [review]

**Review by syedomarazam — APPROVED** [review]
approved

**Comment by syedomarazam** [comment]
Merging into master

**Pull request merged**